<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Single cell data science - Introduction to single cell data science</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Single cell data science</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lecture-material" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Lecture material</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lecture-material">    
        <li>
    <a class="dropdown-item" href="./intro-singlecell.html" rel="" target="">
 <span class="dropdown-text">Introduction to single cell data science</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./intro-bioc.html" rel="" target="">
 <span class="dropdown-text">Introduction to Bioconductor</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./intro-granges.html" rel="" target="">
 <span class="dropdown-text">Introduction to GRanges</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#part-1" id="toc-part-1" class="nav-link active" data-scroll-target="#part-1">Part 1</a>
  <ul class="collapse">
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a></li>
  <li><a href="#materials" id="toc-materials" class="nav-link" data-scroll-target="#materials">Materials</a></li>
  </ul></li>
  <li><a href="#part-2" id="toc-part-2" class="nav-link" data-scroll-target="#part-2">Part 2</a>
  <ul class="collapse">
  <li><a href="#learning-objectives-1" id="toc-learning-objectives-1" class="nav-link" data-scroll-target="#learning-objectives-1">Learning objectives</a></li>
  <li><a href="#palmer-penguins" id="toc-palmer-penguins" class="nav-link" data-scroll-target="#palmer-penguins">Palmer penguins</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization">Visualization</a></li>
  <li><a href="#dimensionality-reduction" id="toc-dimensionality-reduction" class="nav-link" data-scroll-target="#dimensionality-reduction">Dimensionality reduction</a></li>
  </ul></li>
  <li><a href="#part-3" id="toc-part-3" class="nav-link" data-scroll-target="#part-3">Part 3</a>
  <ul class="collapse">
  <li><a href="#learning-objectives-2" id="toc-learning-objectives-2" class="nav-link" data-scroll-target="#learning-objectives-2">Learning objectives</a></li>
  <li><a href="#overview" id="toc-overview" class="nav-link" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#singlecellexperiment-class" id="toc-singlecellexperiment-class" class="nav-link" data-scroll-target="#singlecellexperiment-class"><code>SingleCellExperiment</code> Class</a>
  <ul class="collapse">
  <li><a href="#singlecellexperiment-example" id="toc-singlecellexperiment-example" class="nav-link" data-scroll-target="#singlecellexperiment-example"><code>SingleCellExperiment</code> Example</a></li>
  </ul></li>
  <li><a href="#a-typical-single-cell-workflow" id="toc-a-typical-single-cell-workflow" class="nav-link" data-scroll-target="#a-typical-single-cell-workflow">A typical single-cell workflow</a></li>
  <li><a href="#quick-start-simple" id="toc-quick-start-simple" class="nav-link" data-scroll-target="#quick-start-simple">Quick start (simple)</a></li>
  <li><a href="#for-more-information" id="toc-for-more-information" class="nav-link" data-scroll-target="#for-more-information">For more information</a></li>
  </ul></li>
  <li><a href="#part-4" id="toc-part-4" class="nav-link" data-scroll-target="#part-4">Part 4</a>
  <ul class="collapse">
  <li><a href="#overview-1" id="toc-overview-1" class="nav-link" data-scroll-target="#overview-1">Overview</a></li>
  </ul></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Introduction to single cell data science</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="part-1" class="level1">
<h1>Part 1</h1>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning objectives</h2>
<ol type="1">
<li>Learn what are genomic data and how are they generated?</li>
<li>What is a <code>GRanges</code> object?</li>
<li>What are some advantages of applying tidy principles to genomic data?</li>
<li>What is a <code>SingleCellExperiment</code> object?</li>
</ol>
</section>
<section id="materials" class="level2">
<h2 class="anchored" data-anchor-id="materials">Materials</h2>
<p>We will go through the slides available here:</p>
<ul>
<li><a href="https://docs.google.com/presentation/d/1VFL4OxK44Q7b1c-Vk9PDnroT-R5_aQP2LRfJaHXQedo/edit?usp=sharing" class="uri">https://docs.google.com/presentation/d/1VFL4OxK44Q7b1c-Vk9PDnroT-R5_aQP2LRfJaHXQedo/edit?usp=sharing</a></li>
</ul>
</section>
</section>
<section id="part-2" class="level1">
<h1>Part 2</h1>
<section id="learning-objectives-1" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives-1">Learning objectives</h2>
<ol type="1">
<li>Remind ourselves of basics of <code>tidyverse</code> functions (mostly <code>dplyr</code> and <code>ggplot2</code>).</li>
<li>Introduce how to preprocess data and to perform principal components analysis (PCA), a widely used method for dimensionality reduction.</li>
</ol>
</section>
<section id="palmer-penguins" class="level2">
<h2 class="anchored" data-anchor-id="palmer-penguins">Palmer penguins</h2>
<p>In this section, we give an example of using <code>tidyverse</code> functions with the <code>palmerpenguins</code> dataset available as a CRAN package.</p>
<div class="cell" data-layout-align="center" data-show="true" data-fig.caption="Palmer penguins">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="figures/lter_penguins.png" class="img-fluid figure-img" width="780"></p>
</figure>
</div>
</div>
</div>
<p>[<strong>Source</strong>: <a href="https://github.com/allisonhorst/stats-illustrations">Artwork by Allison Horst</a>]</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(palmerpenguins) <span class="co"># penguins!</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ggplot2) <span class="co"># "grammar of graphics" plots</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(dplyr) <span class="co"># data pliers</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can select first three rows and two columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(species, island) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  species island   
  &lt;fct&gt;   &lt;fct&gt;    
1 Adelie  Torgersen
2 Adelie  Torgersen
3 Adelie  Torgersen</code></pre>
</div>
</div>
<p>And estimate the average body mass in grams using <code>summarize()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">ave_mass =</span> <span class="fu">mean</span>(body_mass_g, <span class="at">na.rm=</span><span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  ave_mass
     &lt;dbl&gt;
1    4202.</code></pre>
</div>
</div>
<p>Rearranging how we omit NA values, we get the same:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>() <span class="sc">|&gt;</span> </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">ave_mass =</span> <span class="fu">mean</span>(body_mass_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  ave_mass
     &lt;dbl&gt;
1    4207.</code></pre>
</div>
</div>
<p>Or we can also drop NA values entirely and modify our original dataset</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="ot">&lt;-</span> penguins <span class="sc">|&gt;</span> </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">ave_mass =</span> <span class="fu">mean</span>(body_mass_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
  ave_mass
     &lt;dbl&gt;
1    4207.</code></pre>
</div>
</div>
<p>A powerful paradigm is to first group and then summarize:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(species, island) <span class="sc">|&gt;</span> </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">ave_mass =</span> <span class="fu">mean</span>(body_mass_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'species'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
# Groups:   species [3]
  species   island    ave_mass
  &lt;fct&gt;     &lt;fct&gt;        &lt;dbl&gt;
1 Adelie    Biscoe       3710.
2 Adelie    Dream        3701.
3 Adelie    Torgersen    3709.
4 Chinstrap Dream        3733.
5 Gentoo    Biscoe       5092.</code></pre>
</div>
</div>
</section>
<section id="visualization" class="level2">
<h2 class="anchored" data-anchor-id="visualization">Visualization</h2>
<p>These data can also be piped into functions for data visualization:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(species, body_mass_g)) <span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill=</span>species))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="intro-singlecell_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="dimensionality-reduction" class="level2">
<h2 class="anchored" data-anchor-id="dimensionality-reduction">Dimensionality reduction</h2>
<p>Next, we will work towards using dimensionality reduction to visualize the penguins in the principal components space.</p>
<p>For the four columns that contain <code>mm</code> or <code>mass</code>, we apply <code>scale()</code>, which centers and scales each variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">contains</span>(<span class="fu">c</span>(<span class="st">"mm"</span>,<span class="st">"mass"</span>)), scale))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 333 × 8
   species island    bill_length_mm[,1] bill_depth_mm[,1] flipper_length_mm[,1]
   &lt;fct&gt;   &lt;fct&gt;                  &lt;dbl&gt;             &lt;dbl&gt;                 &lt;dbl&gt;
 1 Adelie  Torgersen             -0.895             0.780                -1.42 
 2 Adelie  Torgersen             -0.822             0.119                -1.07 
 3 Adelie  Torgersen             -0.675             0.424                -0.426
 4 Adelie  Torgersen             -1.33              1.08                 -0.568
 5 Adelie  Torgersen             -0.858             1.74                 -0.782
 6 Adelie  Torgersen             -0.931             0.323                -1.42 
 7 Adelie  Torgersen             -0.876             1.24                 -0.426
 8 Adelie  Torgersen             -0.529             0.221                -1.35 
 9 Adelie  Torgersen             -0.986             2.05                 -0.711
10 Adelie  Torgersen             -1.72              2.00                 -0.212
# ℹ 323 more rows
# ℹ 3 more variables: body_mass_g &lt;dbl[,1]&gt;, sex &lt;fct&gt;, year &lt;int&gt;</code></pre>
</div>
</div>
<p>Visualizing the scaled data (i.e.&nbsp;y-axis has changed, but distribution is the same).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">contains</span>(<span class="fu">c</span>(<span class="st">"mm"</span>,<span class="st">"mass"</span>)), scale)) <span class="sc">|&gt;</span> </span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(species, body_mass_g)) <span class="sc">+</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill=</span>species))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="intro-singlecell_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s create a new matrix called <code>scaled</code> with just the four columns scaled.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>scaled <span class="ot">&lt;-</span> penguins <span class="sc">|&gt;</span> </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">contains</span>(<span class="fu">c</span>(<span class="st">"mm"</span>,<span class="st">"mass"</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), scale))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>scaled</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 333 × 4
   bill_length_mm[,1] bill_depth_mm[,1] flipper_length_mm[,1] body_mass_g[,1]
                &lt;dbl&gt;             &lt;dbl&gt;                 &lt;dbl&gt;           &lt;dbl&gt;
 1             -0.895             0.780                -1.42           -0.568
 2             -0.822             0.119                -1.07           -0.506
 3             -0.675             0.424                -0.426          -1.19 
 4             -1.33              1.08                 -0.568          -0.940
 5             -0.858             1.74                 -0.782          -0.692
 6             -0.931             0.323                -1.42           -0.723
 7             -0.876             1.24                 -0.426           0.581
 8             -0.529             0.221                -1.35           -1.25 
 9             -0.986             2.05                 -0.711          -0.506
10             -1.72              2.00                 -0.212           0.240
# ℹ 323 more rows</code></pre>
</div>
</div>
<p>Then, we can apply the function <code>prcomp()</code> to calculate the principal components 1 to 4.</p>
<p>Under the hood, the function uses a singular value decomposition of the centered and scaled data matrix (not using <code>eigen</code> on the covariance matrix). Often preferred for numerical accuracy.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>scaled <span class="sc">|&gt;</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Standard deviations (1, .., p=4):
[1] 1.6569115 0.8821095 0.6071594 0.3284579

Rotation (n x k) = (4 x 4):
                         PC1         PC2        PC3        PC4
bill_length_mm     0.4537532 -0.60019490 -0.6424951  0.1451695
bill_depth_mm     -0.3990472 -0.79616951  0.4258004 -0.1599044
flipper_length_mm  0.5768250 -0.00578817  0.2360952 -0.7819837
body_mass_g        0.5496747 -0.07646366  0.5917374  0.5846861</code></pre>
</div>
</div>
<p>Some useful things about output:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> scaled <span class="sc">|&gt;</span> </span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prcomp</span>()</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "sdev"     "rotation" "center"   "scale"    "x"       </code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(pca<span class="sc">$</span>x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 333   4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pca<span class="sc">$</span>x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           PC1         PC2         PC3        PC4
[1,] -1.850808 -0.03202119  0.23454869  0.5276026
[2,] -1.314276  0.44286031  0.02742880  0.4011230
[3,] -1.374537  0.16098821 -0.18940423 -0.5278675
[4,] -1.882455  0.01233268  0.62792772 -0.4721826
[5,] -1.917096 -0.81636958  0.69999797 -0.1961213
[6,] -1.770356  0.36567266 -0.02841769  0.5046092</code></pre>
</div>
</div>
<p>Visualize</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>penguins <span class="sc">|&gt;</span> </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(pca<span class="sc">$</span>x) <span class="sc">|&gt;</span> </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(PC1, PC2, <span class="at">color=</span>species)) <span class="sc">+</span> </span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="intro-singlecell_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="part-3" class="level1">
<h1>Part 3</h1>
<p>In this next section, we move into the analysis of single-cell transcriptomics data with a goal of learning the basic steps for data wrangling these data to calculate top principal components where each point is a <em>cell</em> (instead of a <em>penguin</em>).</p>
<section id="learning-objectives-2" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives-2">Learning objectives</h2>
<ol type="1">
<li>Be able to create a single-cell count matrix and read it into R</li>
<li>Recognize and define the <code>SingleCellExperiment</code> S4 class in R/Bioconductor to store single-cell data</li>
<li>Be able to describe a standard workflow for analyzing single-cell data</li>
<li>Be able to run code for a standard workflow starting from loading a <code>SingleCellExperiment</code> in R and identifying clusters.</li>
</ol>
</section>
<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>NGS data from scRNA-seq experiments must be converted into a matrix of expression values.</p>
<p>This is <strong>usually a count matrix</strong> containing the number of reads (or UMIs) mapped to each gene (row) in each cell (column). Once this quantification is complete, we can proceed with our downstream statistical analyses in R.</p>
<p>Constructing a count matrix from raw scRNA-seq data requires some thought as the term “single-cell RNA-seq” encompasses a variety of different experimental protocols.</p>
</section>
<section id="singlecellexperiment-class" class="level2">
<h2 class="anchored" data-anchor-id="singlecellexperiment-class"><code>SingleCellExperiment</code> Class</h2>
<p>One of the <strong>main strengths</strong> of the Bioconductor project lies in the use of <strong>a common data infrastructure</strong> that powers interoperability across packages.</p>
<p>Users should be able to analyze their data using functions from different Bioconductor packages without the need to convert between formats.</p>
<p>To this end, the <code>SingleCellExperiment</code> class (from the <code>SingleCellExperiment</code> package) serves as the common currency for data exchange across 70+ single-cell-related Bioconductor packages.</p>
<p>This class implements a data structure that stores all aspects of our single-cell data - gene-by-cell expression data, per-cell metadata and per-gene annotation - and manipulate them in a synchronized manner.</p>
<div class="cell" data-layout-align="center" data-show="true" data-fig.caption="Overview of the structure of the `SingleCellExperiment` class. Each row of the assays corresponds to a row of the `rowData` (pink shading), while each column of the assays corresponds to a column of the `colData` and `reducedDims` (yellow shading).">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="figures/SingleCellExperiment.png" class="img-fluid figure-img" width="780"></p>
</figure>
</div>
</div>
</div>
<p>[<a href="https://doi.org/10.1101/590562">source</a>]</p>
<ul>
<li>Each piece of (meta)data in the <code>SingleCellExperiment</code> is <strong>represented by a separate “slot”</strong>.</li>
</ul>
<p>(This terminology comes from the S4 class system, but that’s not important right now).</p>
<p>If we imagine the <code>SingleCellExperiment</code> object to be a cargo ship, the <strong>slots can be thought of as individual cargo boxes with different contents</strong>, e.g., certain slots expect numeric matrices whereas others may expect data frames.</p>
<p>If you want to know more about the available slots, their expected formats, and how we can interact with them, check out this <a href="https://bioconductor.org/books/3.15/OSCA.intro/the-singlecellexperiment-class.html">chapter</a>.</p>
<section id="singlecellexperiment-example" class="level3">
<h3 class="anchored" data-anchor-id="singlecellexperiment-example"><code>SingleCellExperiment</code> Example</h3>
<p>Let’s show you what a <code>SingleCellExperiment</code> (or <code>sce</code> for short) looks like.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sce</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 20006 3005 
metadata(0):
assays(1): counts
rownames(20006): Tspan12 Tshz1 ... mt-Rnr1 mt-Nd4l
rowData names(1): featureType
colnames(3005): 1772071015_C02 1772071017_G12 ... 1772066098_A12
  1772058148_F03
colData names(10): tissue group # ... level1class level2class
reducedDimNames(0):
mainExpName: endogenous
altExpNames(2): ERCC repeat</code></pre>
</div>
</div>
<p>This <code>SingleCellExperiment</code> object has 20006 genes and 3005 cells.</p>
<p>We can pull out the counts matrix with the <code>counts()</code> function and the corresponding <code>rowData()</code> and <code>colData()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">counts</span>(sce)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         1772071015_C02 1772071017_G12 1772071017_A05 1772071014_B06
Tspan12               0              0              0              3
Tshz1                 3              1              0              2
Fnbp1l                3              1              6              4
Adamts15              0              0              0              0
Cldn12                1              1              1              0
         1772067065_H06
Tspan12               0
Tshz1                 2
Fnbp1l                1
Adamts15              0
Cldn12                0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 20006 rows and 1 column
         featureType
         &lt;character&gt;
Tspan12   endogenous
Tshz1     endogenous
Fnbp1l    endogenous
Adamts15  endogenous
Cldn12    endogenous
...              ...
mt-Co2          mito
mt-Co1          mito
mt-Rnr2         mito
mt-Rnr1         mito
mt-Nd4l         mito</code></pre>
</div>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 3005 rows and 10 columns
                       tissue   group # total mRNA mol      well       sex
                  &lt;character&gt; &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
1772071015_C02       sscortex         1           1221         3         3
1772071017_G12       sscortex         1           1231        95         1
1772071017_A05       sscortex         1           1652        27         1
1772071014_B06       sscortex         1           1696        37         3
1772067065_H06       sscortex         1           1219        43         3
...                       ...       ...            ...       ...       ...
1772067059_B04 ca1hippocampus         9           1997        19         1
1772066097_D04 ca1hippocampus         9           1415        21         1
1772063068_D01       sscortex         9           1876        34         3
1772066098_A12 ca1hippocampus         9           1546        88         1
1772058148_F03       sscortex         9           1970        15         3
                     age  diameter        cell_id       level1class level2class
               &lt;numeric&gt; &lt;numeric&gt;    &lt;character&gt;       &lt;character&gt; &lt;character&gt;
1772071015_C02         2         1 1772071015_C02      interneurons       Int10
1772071017_G12         1       353 1772071017_G12      interneurons       Int10
1772071017_A05         1        13 1772071017_A05      interneurons        Int6
1772071014_B06         2        19 1772071014_B06      interneurons       Int10
1772067065_H06         6        12 1772067065_H06      interneurons        Int9
...                  ...       ...            ...               ...         ...
1772067059_B04         4       382 1772067059_B04 endothelial-mural       Peric
1772066097_D04         7        12 1772066097_D04 endothelial-mural        Vsmc
1772063068_D01         7       268 1772063068_D01 endothelial-mural        Vsmc
1772066098_A12         7       324 1772066098_A12 endothelial-mural        Vsmc
1772058148_F03         7         6 1772058148_F03 endothelial-mural        Vsmc</code></pre>
</div>
</div>
</section>
</section>
<section id="a-typical-single-cell-workflow" class="level2">
<h2 class="anchored" data-anchor-id="a-typical-single-cell-workflow">A typical single-cell workflow</h2>
<p>Here, we provide an overview of the framework of a typical scRNA-seq analysis workflow:</p>
<div class="cell" data-layout-align="center" data-show="true" data-fig.caption="Schematic of a typical scRNA-seq analysis workflow. Each stage (separated by dashed lines) consists of a number of specific steps, many of which operate on and modify a SingleCellExperiment instance.">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="figures/workflow.png" class="img-fluid figure-img" width="780"></p>
</figure>
</div>
</div>
</div>
<p>In the simplest case, the workflow has the following form:</p>
<ol type="1">
<li>We compute <strong>quality control metrics</strong> to remove low-quality cells that would interfere with downstream analyses. These cells may have been damaged during processing or may not have been fully captured by the sequencing protocol. Common metrics includes the total counts per cell, the proportion of spike-in or mitochondrial reads and the number of detected features.</li>
<li>We convert the counts into <strong>normalized expression values</strong> to eliminate cell-specific biases (e.g., in capture efficiency). This allows us to perform explicit comparisons across cells in downstream steps like clustering. We also apply a transformation, typically log, to adjust for the mean-variance relationship.</li>
<li>We perform <strong>feature selection to pick a subset of interesting features</strong> for downstream analysis. This is done by modelling the variance across cells for each gene and retaining genes that are highly variable. The aim is to reduce computational overhead and noise from uninteresting genes.</li>
<li>We apply <strong>dimensionality reduction to compact the data</strong> and further reduce noise. Principal components analysis is typically used to obtain an initial low-rank representation for more computational work, followed by more aggressive methods like <span class="math inline">\(t\)</span>-stochastic neighbor embedding for visualization purposes.</li>
<li>We <strong>cluster cells into groups</strong> according to similarities in their (normalized) expression profiles. This aims to obtain groupings that serve as empirical proxies for distinct biological states. We typically interpret these groupings by identifying differentially expressed marker genes between clusters.</li>
</ol>
</section>
<section id="quick-start-simple" class="level2">
<h2 class="anchored" data-anchor-id="quick-start-simple">Quick start (simple)</h2>
<p>Here, we use the a droplet-based retina dataset from Macosko et al.&nbsp;(2015), provided in the <code>scRNAseq</code> package. This starts from a count matrix and finishes with clusters in preparation for biological interpretation. We also demonstrate how to identify differentially expressed genes between the clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scRNAseq)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">MacoskoRetinaData</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?scRNAseq and browseVignettes('scRNAseq') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?scRNAseq and browseVignettes('scRNAseq') for documentation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Quality control (using mitochondrial genes).</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: scuttle</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>is.mito <span class="ot">&lt;-</span> <span class="fu">grepl</span>(<span class="st">"^MT-"</span>, <span class="fu">rownames</span>(sce)) <span class="co"># find mitochondrial genes</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>qcstats <span class="ot">&lt;-</span> <span class="fu">perCellQCMetrics</span>(sce, <span class="at">subsets=</span><span class="fu">list</span>(<span class="at">Mito=</span>is.mito)) <span class="co"># calculate QC metrics </span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>filtered <span class="ot">&lt;-</span> <span class="fu">quickPerCellQC</span>(qcstats, <span class="at">percent_subsets=</span><span class="st">"subsets_Mito_percent"</span>) <span class="co"># filter base on QC metrics</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[, <span class="sc">!</span>filtered<span class="sc">$</span>discard] <span class="co"># subset object for post-QC analyses</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalization.</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(sce)</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Feature selection (which genes are most important)? </span></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>dec.retina <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce)</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>hvg <span class="ot">&lt;-</span> <span class="fu">getTopHVGs</span>(dec.retina, <span class="at">prop=</span><span class="fl">0.1</span>)</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualizing the fit:</span></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>fit.retina <span class="ot">&lt;-</span> <span class="fu">metadata</span>(dec.retina)</span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(fit.retina<span class="sc">$</span>mean, fit.retina<span class="sc">$</span>var, <span class="at">xlab=</span><span class="st">"Mean of log-expression"</span>,</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab=</span><span class="st">"Variance of log-expression"</span>)</span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a><span class="fu">curve</span>(fit.retina<span class="sc">$</span><span class="fu">trend</span>(x), <span class="at">col=</span><span class="st">"dodgerblue"</span>, <span class="at">add=</span><span class="cn">TRUE</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="intro-singlecell_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">ncomponents=</span><span class="dv">25</span>, <span class="at">subset_row=</span>hvg)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Clustering.</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bluster)</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colLabels</span>(sce) <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce, <span class="at">use.dimred=</span><span class="st">'PCA'</span>,</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">BLUSPARAM=</span><span class="fu">NNGraphParam</span>(<span class="at">cluster.fun=</span><span class="st">"louvain"</span>))  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualization of PC1 vs PC2</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by=</span><span class="st">"label"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="intro-singlecell_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP visualization </span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runUMAP</span>(sce, <span class="at">dimred =</span> <span class="st">'PCA'</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotUMAP</span>(sce, <span class="at">colour_by=</span><span class="st">"label"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="intro-singlecell_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">UMAP plot of the retina dataset, where each point is a cell and is colored by the assigned cluster identity.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="for-more-information" class="level2">
<h2 class="anchored" data-anchor-id="for-more-information">For more information</h2>
<p>If you could like to know more, check out this book:</p>
<ul>
<li><a href="https://bioconductor.org/books/release/OSCA/" class="uri">https://bioconductor.org/books/release/OSCA/</a></li>
</ul>
</section>
</section>
<section id="part-4" class="level1">
<h1>Part 4</h1>
<p>Last, we will briefly highlight the <code>tidySingleCellExperiment</code> package (<a href="https://www.bioconductor.org/packages/SingleCellExperiment" class="uri">https://www.bioconductor.org/packages/SingleCellExperiment</a>) that allows you to leverage functions from <code>dplyr</code>, <code>tidyr</code>, and <code>ggplot2</code> with a <code>SingleCellExperiment</code> object.</p>
<p>This package is part of a large set of packages called <code>tidyomics</code>, which bridges the two worlds of tidy principles in R with Bioconductor.</p>
<p>It is similar to <code>tidyverse</code> in that it installs a series of packages that are designed to work with different types of objects in Bioconductor. Specifically, it installs:</p>
<ul>
<li><code>plyranges</code> = for data with <code>GRanges</code></li>
<li><code>tidybulk</code> = for bulk RNA-sequencing data</li>
<li><code>tidySummarizedExperiment</code> = for data in a <code>SummarizedExperiment</code> object</li>
<li><code>tidySingleCellExperiment</code> = for data in a <code>SingleCellExperiment</code> object</li>
<li><code>tidySpatialExperiment</code> = for data in a <code>SpatialExperiment</code> object</li>
<li><code>tidyseurat</code> = to bring Seurat to the tidyverse</li>
<li><code>nullranges</code> = to generate of sets of <code>GRranges</code> representing the null hypothesis</li>
</ul>
<p>For more information:</p>
<ul>
<li>manuscript: <a href="https://doi.org/10.1101/2023.09.10.557072" class="uri">https://doi.org/10.1101/2023.09.10.557072</a></li>
<li>to install <code>tidyomics</code>: <a href="https://github.com/tidyomics/tidyomics" class="uri">https://github.com/tidyomics/tidyomics</a></li>
</ul>
<div class="sourceCode" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"tidyomics/tidyomics"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="overview-1" class="level2">
<h2 class="anchored" data-anchor-id="overview-1">Overview</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidySummarizedExperiment) <span class="co"># allow dplyr</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>sce <span class="sc">|&gt;</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">addPerCellQC</span>(<span class="at">subsets=</span><span class="fu">list</span>(<span class="at">Mito=</span>is.mito)) <span class="sc">|&gt;</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colData</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 45877 rows and 10 columns
                        cell.id   cluster sizeFactor    label       sum
                    &lt;character&gt; &lt;integer&gt;  &lt;numeric&gt; &lt;factor&gt; &lt;numeric&gt;
r1_GGCCGCAGTCCG r1_GGCCGCAGTCCG         2    28.0131        1     37487
r1_CTTGTGCGGGAA r1_CTTGTGCGGGAA         2    23.9479        1     32047
r1_GCGCAACTGCTC r1_GCGCAACTGCTC         2    21.0343        1     28148
r1_GATTGGGAGGCA r1_GATTGGGAGGCA         2    15.2197        1     20367
r1_CCTCCTAGTTGG r1_CCTCCTAGTTGG        NA    14.6167        1     19560
...                         ...       ...        ...      ...       ...
p1_TCAAAAGCCGGG p1_TCAAAAGCCGGG        24   0.610523       13       817
p1_ATTAAGTTCCAA p1_ATTAAGTTCCAA        34   0.610523       5        817
p1_CTGTCTGAGACC p1_CTGTCTGAGACC         2   0.610523       1        817
p1_TAACGCGCTCCT p1_TAACGCGCTCCT        24   0.609776       11       816
p1_ATTCTTGTTCTT p1_ATTCTTGTTCTT        24   0.609776       8        816
                 detected subsets_Mito_sum subsets_Mito_detected
                &lt;integer&gt;        &lt;numeric&gt;             &lt;integer&gt;
r1_GGCCGCAGTCCG      7243              427                    14
r1_CTTGTGCGGGAA      6933              503                    15
r1_GCGCAACTGCTC      6397              460                    13
r1_GATTGGGAGGCA      5740              326                    11
r1_CCTCCTAGTTGG      5779              264                     9
...                   ...              ...                   ...
p1_TCAAAAGCCGGG       537               13                     4
p1_ATTAAGTTCCAA       574               10                     5
p1_CTGTCTGAGACC       637               24                     7
p1_TAACGCGCTCCT       488               27                     5
p1_ATTCTTGTTCTT       484               16                     4
                subsets_Mito_percent     total
                           &lt;numeric&gt; &lt;numeric&gt;
r1_GGCCGCAGTCCG              1.13906     37487
r1_CTTGTGCGGGAA              1.56957     32047
r1_GCGCAACTGCTC              1.63422     28148
r1_GATTGGGAGGCA              1.60063     20367
r1_CCTCCTAGTTGG              1.34969     19560
...                              ...       ...
p1_TCAAAAGCCGGG              1.59119       817
p1_ATTAAGTTCCAA              1.22399       817
p1_CTGTCTGAGACC              2.93758       817
p1_TAACGCGCTCCT              3.30882       816
p1_ATTCTTGTTCTT              1.96078       816</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify variable genes with scran</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>variable_genes <span class="ot">&lt;-</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>    sce <span class="sc">%&gt;%</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">modelGeneVar</span>() <span class="sc">%&gt;%</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getTopHVGs</span>(<span class="at">prop=</span><span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform PCA with scater</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>retina_pca <span class="ot">&lt;-</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>    sce <span class="sc">%&gt;%</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">runPCA</span>(<span class="at">subset_row=</span>variable_genes)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(retina_pca, <span class="at">colour_by=</span><span class="st">"label"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="intro-singlecell_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="session-info" class="level1">
<h1>Session Info</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.1 (2023-06-16)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] tidySummarizedExperiment_1.10.0 bluster_1.10.0                 
 [3] scran_1.28.2                    scater_1.28.0                  
 [5] scuttle_1.9.4                   scRNAseq_2.14.0                
 [7] SingleCellExperiment_1.22.0     SummarizedExperiment_1.30.2    
 [9] Biobase_2.60.0                  GenomicRanges_1.52.0           
[11] GenomeInfoDb_1.36.3             IRanges_2.34.1                 
[13] S4Vectors_0.38.1                BiocGenerics_0.46.0            
[15] MatrixGenerics_1.12.3           matrixStats_1.0.0              
[17] dplyr_1.1.3                     ggplot2_3.4.3                  
[19] palmerpenguins_0.1.1           

loaded via a namespace (and not attached):
  [1] rstudioapi_0.15.0             jsonlite_1.8.7               
  [3] magrittr_2.0.3                ggbeeswarm_0.7.2             
  [5] GenomicFeatures_1.52.2        farver_2.1.1                 
  [7] rmarkdown_2.24                BiocIO_1.10.0                
  [9] zlibbioc_1.46.0               vctrs_0.6.3                  
 [11] memoise_2.0.1                 Rsamtools_2.16.0             
 [13] DelayedMatrixStats_1.22.6     RCurl_1.98-1.12              
 [15] htmltools_0.5.6               S4Arrays_1.0.6               
 [17] progress_1.2.2                AnnotationHub_3.8.0          
 [19] curl_5.0.2                    BiocNeighbors_1.18.0         
 [21] htmlwidgets_1.6.2             plotly_4.10.2                
 [23] cachem_1.0.8                  GenomicAlignments_1.36.0     
 [25] igraph_1.5.1                  mime_0.12                    
 [27] lifecycle_1.0.3               pkgconfig_2.0.3              
 [29] rsvd_1.0.5                    Matrix_1.6-1                 
 [31] R6_2.5.1                      fastmap_1.1.1                
 [33] GenomeInfoDbData_1.2.10       shiny_1.7.5                  
 [35] digest_0.6.33                 colorspace_2.1-0             
 [37] AnnotationDbi_1.62.2          dqrng_0.3.1                  
 [39] irlba_2.3.5.1                 ExperimentHub_2.8.1          
 [41] RSQLite_2.3.1                 beachmat_2.16.0              
 [43] filelock_1.0.2                labeling_0.4.3               
 [45] fansi_1.0.4                   httr_1.4.7                   
 [47] abind_1.4-5                   compiler_4.3.1               
 [49] bit64_4.0.5                   withr_2.5.0                  
 [51] BiocParallel_1.34.2           viridis_0.6.4                
 [53] DBI_1.1.3                     biomaRt_2.56.1               
 [55] rappdirs_0.3.3                DelayedArray_0.26.7          
 [57] rjson_0.2.21                  tools_4.3.1                  
 [59] vipor_0.4.5                   beeswarm_0.4.0               
 [61] interactiveDisplayBase_1.38.0 httpuv_1.6.11                
 [63] glue_1.6.2                    restfulr_0.0.15              
 [65] promises_1.2.1                grid_4.3.1                   
 [67] cluster_2.1.4                 generics_0.1.3               
 [69] gtable_0.3.4                  tidyr_1.3.0                  
 [71] ensembldb_2.24.0              data.table_1.14.8            
 [73] hms_1.1.3                     metapod_1.7.0                
 [75] BiocSingular_1.16.0           ScaledMatrix_1.8.1           
 [77] xml2_1.3.5                    utf8_1.2.3                   
 [79] XVector_0.40.0                RcppAnnoy_0.0.21             
 [81] ggrepel_0.9.3                 BiocVersion_3.17.1           
 [83] pillar_1.9.0                  stringr_1.5.0                
 [85] limma_3.56.2                  later_1.3.1                  
 [87] BiocFileCache_2.8.0           lattice_0.21-8               
 [89] rtracklayer_1.60.1            bit_4.0.5                    
 [91] tidyselect_1.2.0              locfit_1.5-9.8               
 [93] Biostrings_2.68.1             knitr_1.44                   
 [95] gridExtra_2.3                 ProtGenerics_1.32.0          
 [97] edgeR_3.42.4                  xfun_0.40                    
 [99] statmod_1.5.0                 stringi_1.7.12               
[101] lazyeval_0.2.2                yaml_2.3.7                   
[103] evaluate_0.21                 codetools_0.2-19             
[105] tibble_3.2.1                  BiocManager_1.30.22          
[107] cli_3.6.1                     uwot_0.1.16                  
[109] xtable_1.8-4                  munsell_0.5.0                
[111] Rcpp_1.0.11                   dbplyr_2.3.3                 
[113] png_0.1-8                     XML_3.99-0.14                
[115] parallel_4.3.1                ellipsis_0.3.2               
[117] blob_1.2.4                    prettyunits_1.1.1            
[119] AnnotationFilter_1.24.0       sparseMatrixStats_1.12.2     
[121] bitops_1.0-7                  viridisLite_0.4.2            
[123] scales_1.2.1                  purrr_1.0.2                  
[125] crayon_1.5.2                  rlang_1.1.1                  
[127] KEGGREST_1.40.0              </code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>